client/gutenberg/editor/init.js

```diff
@@ -60,6 +63,18 @@ const addResetToRegistry = registry => {
 	};
 };
 
+const registerGutenbergBlocks = registerCoreBlocks =>
+	waitForData( {
+		serverBlocksAttributes: () => requestGutenbergServerBlocksAttributes(),
+	} ).then( ( { serverBlocksAttributes } ) => {
+		if ( serverBlocksAttributes.data ) {
+			debug( 'Registering server-defined blocks attributes' );
+			unstable__bootstrapServerSideBlockDefinitions( serverBlocksAttributes.data );
+		}
+		debug( 'Registering core blocks' );
+		registerCoreBlocks();
+	} );
```

Comment:
I don't know if we can find a good way to split it, but I'm noting that we seem to be introducing another API call serially here when I think we could be happy making it in parallel to other calls that prepare the editor.

if you look at [the EditorLoader](https://github.com/Automattic/wp-calypso/blob/c5f364d218c0a9cef188f6a3cdc952d312147bcc/client/gutenberg/editor/controller.js#L188) you can see where we list parallel-executed promises. it might end up a little complicated if we have to pass in the result of that API call to `init()` but it can be done and it would prevent another loading state.

I didn't see anything inside of `init` that was async. we might be able to rearrange `makeEditor` so that instead of calling the `require()` promise it gets that as its input.

```js
	const EditorLoader = asyncLoader( {
		promises: {
			editorModule: import( './init' ),
			serverBlocks: requestEm(),
			translations: loadTranslations( context.store ),
			blockAvailability: loadGutenbergBlockAvailability( context.store ),
		},
		loading: () => <Placeholder />,
		success: ( { editorModule: { init }, serverBlocks } ) => React.createElement( makeEditor( init, serverBlocks ) ),
		failure: () => <div>Couldn't load everything - try hitting reload in your browser…</div>,
	} );
```

That would start the load of the server blocks as soon as we start loading the editor code.

Also, the server block definitions are inside of the `once()` and I wonder if that will need to be evaluated in context of changing sites.

Reply:
Hmm, are we overthinking this maybe? This definition should remain the same for simple sites on wpcom, so I think we can maybe even load this once in the app skeleton or layout. No Jetpack sites use Gutenlypso yet.

Reply:
I wasn't really looking at the broader problem @gwwar so @Copons yeah please defer to this comment above.

on that note though _if_ we end up making the API call I was mostly concerned about introducing the async task in the serial hot path vs. side loading it in parallel, even if it's only loaded once

Reply:
We also wouldn't support non-whitelisted blocks in Calypso, so I suppose a sublist for Jetpack sites wouldn't hurt either

Reply:
My reasonings for running this in the `init` instead of in the controller async loader are:
- Keep the entire blocks registration logic in one single place.
- Make sure to only run the server-side attributes registration once.

Although, I agree that the controller's `asyncLoader` is a much better fit for `waitForData`.

Reply:
Sorry, I've somehow forgot to complete the thought.

I agree that `asyncLoader` is the right fit for `waitForData` functions.
But, since this is now only about Core blocks and Core blocks never change and are the same for all sites, it's also optimal to run this request only once.

I know that subsequent requests would always hit the cache, but they'd still be unnecessary.